src/Microsoft.ML.LightGBM/LightGbmArguments.cs

```diff
@@ -371,6 +371,15 @@ public enum EvalMetricType
         [TlcModule.SweepableDiscreteParam("CatL2", new object[] { 0.1, 0.5, 1, 5, 10 })]
         public double CatL2 = 10;
 
+        [Argument(ArgumentType.Multiple,
+            HelpText = "Sets the constraints for monotonic features. This is a 0 based index for each feature in " +
+            "the features column. A keyword of 'pos' for positive constraint or 'neg' for negative constraint is " +
+            "specified followed by a range. For example, pos:0-2 neg:3,5 will apply a positive constraint to the " +
+            "first three features and a negative constraint to the 4th and 6th feature. If feature index is not specified, " +
+            "then no constraint will be applied. The keyword of 'pos' or 'neg' without a range will apply the constraint to all features.",
+            ShortName="mc")]
+        public string[] MonotoneConstraints;
```

Comment:
>MonotoneConstraints [](start = 24, length = 19)

Sorry what I bring it on 9th iteration, but just curious, wouldn't it be easier to have two options, one is positive monotone constraints, and other negative with types int[]?

Reply:
or array of Range, or what class we use in textloader to specify range of columns.

---
In reply to: [254096866](https://github.com/dotnet/machinelearning/pull/2330#discussion_r254096866) [](ancestors = 254096866)

Reply:
Using the TextLoader's text-based ranges offers interesting abilities, like "~" and "*". For example setting "pos:3-*" so the user doesn't need to know the total number of slots. 

The current method is interesting as it allows for setting everything positive, then some negative "pos neg:5-10". The gain here is the user doesn't need to know the total number of slots. 

I don't see a shortcut to set positive on all except for a few: "pos neutral:5-10". If you know the total number of slots, you could do "pos:0-4,11-547542".

Reply:
So initially I did not use TextLoader Range as it looked to contain properties specific to Text Loading -- however, your comments resonated with me so I have made some changes:
1) I created a new Range class in Microsoft.ML.Core - this is meant to be a more generic Range class that can be used by other Arguments if a range is needed. Tests were added as well. 
2) I removed MonotonicConstraints as an argument and added MonotonicPositive and MonotonicNegative - these are now Range arrays.
3) I updated tests and baseline files to reflect these changes. 

Overall I like this better as it follows the range format that is already used for text loader. If you want to select all features you can use the * and set the range to be 0-*. Short names for these arguments are mp and mn.

Take a look  - Im curious to what you think. There maybe an opportunity to merge the Microsoft.ML.Core Range and the TextLoader Range, but I think that should be a different discussion.


---
In reply to: [254097097](https://github.com/dotnet/machinelearning/pull/2330#discussion_r254097097) [](ancestors = 254097097,254096866)

Reply:
I didn't expect you to take me seriously on this proposal, I was just thinking out loud.
I'm fine with changes, but I would like @TomFinley to give his opinion regarding new `Range` class, and how it fit in our API story.

Reply:
Thanks @Ivanidzo4ka, it looks this PR fails .netcore30 builds because there is now a System.Range. So I am curious as to what @TomFinley has to say.